components/local-app/pkg/bastion/bastion.go

```diff
@@ -290,3 +493,109 @@ func installSSHAuthorizedKey(ws *Workspace, key string) error {
 
 	return nil
 }
+
+func (b *Bastion) GetTunnelLocalAddr(workspaceID string, port uint32) (string, error) {
+	b.workspacesMu.RLock()
```

Comment:
`defer b.workspacesMu.RULock()` is much preferable here because it makes reasoning about the unlocking behaviour much simpler (don't have to trace all branches manually)